src/main/kotlin/org/rust/ide/inspections/RsVariableMutableInspection.kt

```diff
@@ -0,0 +1,53 @@
+package org.rust.ide.inspections
+
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.psi.PsiElement
+import com.intellij.psi.search.LocalSearchScope
+import com.intellij.psi.search.searches.ReferencesSearch
+import org.rust.ide.annotator.fixes.RemoveMutableFix
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.resolve.ref.RsMethodCallReferenceImpl
+
+class RsVariableMutableInspection : RsLocalInspectionTool() {
+    override fun getDisplayName() = "No mutable required"
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean) =
+        object : RsVisitor() {
+            override fun visitPatBinding(o: RsPatBinding) {
+                if (!o.isMut) return
+                val block = o.parentOfType<RsBlock>() ?: o.parentOfType<RsFunction>() ?: return
+                val macro = block.descendantsOfType<RsMacroExpr>()
```

Comment:
There could be a small optimization: you don't need to check macros if you find a mutable usage. So, this call to `descendantsOfType` (which builds the whole collection of macro calls) should go after the occurrences check to skip it sometimes.

It also can be used the same way as `checkOccurence()` instead of introducing a local variable:

```kotlin
if (block.descendantsOfType<RsMacroExpr>().any { ... }) return
```